libgalois/include/katana/PropertyGraph.h

```diff
@@ -41,49 +56,53 @@ class KATANA_EXPORT GraphTopology {
 
   GraphTopology(
       LargeArray<Edge>&& adj_indices, LargeArray<Node>&& dests) noexcept
-      : adj_indices_(std::move(adj_indices)),
-        dests_(std::move(dests)),
-        out_indices(LargeToArrowArray(adj_indices_)),
-        out_dests(LargeToArrowArray(dests_)) {}
-
-  static std::unique_ptr<GraphTopology> Copy(
-      const GraphTopology& that) noexcept;
-
-  /*
-   * TODO(amber): consider re-enabling these in the near future
-  GraphTopology(GraphTopology&& that):
-    adj_indices_(std::move(that.adj_indices_)),
-    dests_(std::move(that.dests)),
-    out_indices(LargeToArrowArray(adj_indices_)),
-    out_dests(LargeToArrowArray(dests_))
-  {}
-
-  GraphTopology& operator = (GraphTopology&& that) {
+      : adj_indices_(std::move(adj_indices)), dests_(std::move(dests)) {}
+  // out_indices(LargeToArrowArray(adj_indices_)),
+  // out_dests(LargeToArrowArray(dests_)) {}
+
+  static GraphTopology Copy(const GraphTopology& that) noexcept;
+
+  GraphTopology(GraphTopology&& that)
+      : adj_indices_(std::move(that.adj_indices_)),
+        dests_(std::move(that.dests_)) {}
+
+  GraphTopology& operator=(GraphTopology&& that) {
     adj_indices_ = std::move(that.adj_indices_);
     dests_ = std::move(that.dests_);
-    out_indices = LargeToArrowArray(adj_indices_);
-    out_dests = LargeToArrowArray(dests_);
     return *this;
   }
-  */
 
   uint64_t num_nodes() const { return adj_indices_.size(); }
 
   uint64_t num_edges() const { return dests_.size(); }
 
-  bool Equals(const GraphTopology& other) const {
-    return out_indices->Equals(*other.out_indices) &&
-           out_dests->Equals(*other.out_dests);
+  const Edge* adj_data() const noexcept { return &adj_indices_[0]; }
+
+  const Node* dest_data() const noexcept { return &dests_[0]; }
```

Comment:
I disagree with providing silly-old getters as they 1) expose internal type of the array 2) are prone to people (inadvertently) trying to copy, or worse, move the arrays.

Reply:
I'd be happier if we introduce a Buffer type and return a buffer with `const T*` and `a size`, so the user doesn't rely on `num_nodes()` or `num_edges`

Reply:
@amberhassaan you just described an arrow array pretty much. We could just wrap the underlying memory in an arrow array (that does not take ownership of the buffer) and return that.

Reply:
That said, not providing any access to the buffers would be better. Best to just provide abstract accessors so no one can observe the internal data layout.

Reply:
I agree with Arthur. We should either drop these or return them as an `arrow::Array`

Reply:
Yes, but Arrow Arrays come with minor annoyances that add up to anger. Size is int64_t. Don't have `operator[]`. Not sure but seems like I need a `shared_ptr<arrow::Buffer>` to get started with creating the array. Feels too much. I just want a 
```
template <typename T>
Buffer {
  const T* data_;
  size_t size_;
...
};

// and may be
template <typename T>
MutableBuffer {
   T* data_;
   size_t size_;
};
```

Reply:
Can we just remove these accessors then? They totally break the abstraction you were trying to create anyway.

Reply:
Not really. I need a back-door. I can either bring back the GraphTopologyAccess class or I can use these accessors and do a `const_cast` to make it obvious that I'm doing something bad.

Reply:
[This RFC](https://docs.google.com/document/d/16XxEBlbaP7yBSxFPIB_XDRP4CyqxAfSdScR0Q12TFKU/edit?usp=sharing) contains my idea for how to wrap nodes and edges into arrow in the long term. It doesn't address the allocators, only the semantics, to allow us to continue using some of arrow's functionality w.r.t parquet, ipc, compute, etc. The custom classes created there could implement `operator[]`, which is not valid in arrow in general due to nullability.
This PR need not consider that if you don't want it to, but there seems to be enough of a discussion on semantics to merit bringing it up.

Reply:
@amberhassaan Where do you need the back door? Generally needing a back door implies the rest of the interface is incomplete. Maybe there needs to be a more capable constructor or Make function in this class.

Reply:
@arthurp : I need it temporarily. There are functions like `SortEdgesByDest(pg_ptr)` that modify the topology arrays. Then some code in Partitioning (libgluon/Partition.cpp) that modifies the arrays. 
I share the desire of a future without such back doors.

Reply:
@danielmawhirter : I agree that arrow Arrays offer benefit when accessing storage. I'm not convinced yet that they offer a benefit when it comes to in-memory topologies, which I'm hoping we'll have a few more of in the near future. 

Once storage makes use of arrow arrays for topology, it may make sense for in-memory representation to follow in vein, but at least for now, LargeArray (NumaArray) offers numa-aware performance benefits and works well with various places in code, such as computing transposed or sorted topologies, even in some analytics apps, and in libgluon. Most of these places were converting LargeArrays to Arrow Arrays and leaking memory. 

Also, once we figure out num-aware allocators for Arrow Arrays, we should revisit the topology representation.

Reply:
I didn't realize it's temporary. Just leave them as returning pointers and add todos at their declarations and use. Hopefully we can update the other code to use the abstraction soon.

Re: NUMA Aware arrow arrays. I actually know how to do it. I just haven't spent the time. I assumed there was someone else tracking it. See https://arrow.apache.org/docs/cpp/memory.html#memory-pools. It's actually pretty straight forward. I don't think it matters that the allocator uses parallel sections. It's still just an allocator. Obviously we would need to use a different allocator for allocations inside parallel sections than in the main thread. But this is already true, IIUC.

Reply:
@danielmawhirter is aware of these differences and is looking into it. We will need different types: `arrow::Array`, `katana::gstl::ArrowArray`, `katana::NUMAArrowArray` (analogous to `std::vector`, `katana::gstl::Vector`, and `katana::NUMAArray` respectively).

Reply:
I'll step back from this, because I have no understanding why a different class is needed for each of these cases. It seems like they should all the same interface to me. But it's not my area, so I'll leave it to y'all. I just worry that a proliferation of classes which provide extremely similar interfaces, but incompatibly, will cause maintenance problems in the future.

Reply:
For arrow, we will NOT need different classes; just different types. The difference b/w the types will be different allocators. But all of them will NOT provide the same semantics. While `katana::gstl::Vector` can replace all instances of `std::vector`, `katana::NUMAArray` cannot (it's not about performance but about correctness).